qms-tests/sanity/tests/documents/ES-20.spec.ts

```diff
@@ -0,0 +1,100 @@
+  test.beforeEach(async ({ page }) => {
+    await (await page.goto(`${PlatformURI}/${HomepageURI}`))?.finished()
+  })
+
+  test.afterEach(async ({ browser }) => {
+    const contexts = browser.contexts()
+    for (const context of contexts) {
+      await context.close()
+    }
+  })
+
+  test('TESTS-277 - @PDF Author can Generate PDF from a doc with all Reviewers and Approvers signature info displayed', async ({
+    page
+  }) => {
+    await allure.description('Requirement\nUsers need to review the document and review status is displayed in PDF')
+    await allure.tms('TESTS-277', 'https://tracex.hc.engineering/workbench/platform/tracker/TESTS-277')
+    const approveDocument: NewDocument = {
+      template: 'HR (HR)',
+      title: 'Approve document-Test',
+      description: 'Approve document description-Test'
+    }
+    const documentDetails: DocumentDetails = {
+      type: 'HR',
+      category: 'Human Resources',
+      version: 'v0.1',
+      status: DocumentStatus.DRAFT,
+      owner: 'Appleseed John',
+      author: 'Appleseed John'
+    }
+
+    await prepareDocumentStep(page, approveDocument)
+    const documentContentPage = new DocumentContentPage(page)
+    await test.step('2. Add reviewers from team', async () => {
+      await documentContentPage.addReviewersFromTeam()
+      await documentContentPage.sendForReview()
+      await documentContentPage.completeReview()
+    })
+    await test.step('3. Send for Approval', async () => {
+      await documentContentPage.buttonSendForApproval.click()
+      await documentContentPage.fillSelectApproversForm([documentDetails.owner])
+      await documentContentPage.checkDocumentStatus(DocumentStatus.IN_APPROVAL)
+      await documentContentPage.checkDocument({
+        ...documentDetails,
+        status: DocumentStatus.IN_APPROVAL
+      })
+      await documentContentPage.checkCurrentRights(DocumentRights.VIEWING)
+    })
+
+    await test.step('4. Approve document', async () => {
+      await documentContentPage.confirmApproval()
+    })
+
+    await test.step('5. Check the document and status', async () => {
+      await documentContentPage.checkDocumentStatus(DocumentStatus.EFFECTIVE)
+      await documentContentPage.checkDocument({
+        ...documentDetails,
+        status: DocumentStatus.EFFECTIVE,
+        version: 'v0.1'
+      })
+      await documentContentPage.checkCurrentRights(DocumentRights.VIEWING)
+
+      await documentContentPage.openApprovals()
+      const documentApprovalsPage = new DocumentApprovalsPage(page)
+      await documentApprovalsPage.checkSuccessApproval(documentDetails.owner)
+      await attachScreenshot('TESTS-277_approve_document.png', page)
+    })
+    await test.step('6. Show Full screen preview', async () => {
+      await documentContentPage.clickDocumentThreeDots()
+      const pdfPages = new PdfPages(page)
+      await pdfPages.printToPdfClick()
+      await page.waitForTimeout(2000)
```

Comment:
Is it safe to wait 2s for pdf generation? Is there a more reliable way to check that the generation is completed?

Reply:
I tested it and it was working fine with 2s, the problem is that the pdf is inside the iframe and we can't detect it. I can maybe create a function to check if the content of an iframe is generated

Reply:
I'm not sure you can check the content of the PDF viewer inside the iframe but you can probably wait for some iframe events like`domcontentloaded`/`networkidle0` or something which could help to improve the stability of the test and be less dependent from some random factors.

Reply:
that can be a bit risky since in some cases networkidle get stuck, but I can add in case it get stuck than it would be something like a promise race where it will be if the networkidle is done first proceed, if not in certain period then do that timeout


```
export async function waitForNetworIdle (page: Page, timeout = 2000): Promise<void> {
  await Promise.race([page.waitForLoadState('networkidle'), new Promise((resolve) => setTimeout(resolve, timeout))])
}
```

Reply:
It was just an example and if it doesn't work in this case we should not use it. But if it's supposed to go "networkidle" normally I'd use this approach and with the bigger race timeout. Also, we can wait for a combination of events if it helps to achieve more stability for the test.
IMO The main goal would be not to save 2 or even 15 seconds every run but to prevent flakiness of the test as a flaky test causes much more time being wasted later.

Reply:
totally agreed on this one, in my opinion this is currently more than enough. I will try with race condition to see how it behaves since well this is a bit tricky. The tests were passing without an issue with this hardcoded timeouts. I will try to update it to not use them